src/token_bucket.rs

```diff
@@ -11,28 +50,89 @@ pub struct TokenBucket<'a> {
+    /// given entity: no tokens can be consumed in that case, regardless of how
+    /// much time passes or how many attempts are performed.
+    ///
+    /// ```
+    /// use std::time::Duration;
+    /// use youshallnotpass::{TokenBucket, Error};
+    ///
+    /// // create a new bucket that does not allow to consume any tokens
+    /// let bucket = TokenBucket::new(0, Duration::from_secs(60));
+    /// assert!(matches!(bucket.consume(1), Err(Error::Blocked)));
+    /// ```
     pub fn new(limit: usize, interval: Duration) -> Self {
         TokenBucket::with_timer(limit, interval, &Instant::now)
     }
 
+    /// Same as [`TokenBucket::new()`], but allows to override the internal clock,
+    /// which is mainly useful in tests.
     pub(crate) fn with_timer(
         limit: usize,
         interval: Duration,
         clock: &'a (dyn Fn() -> Instant + Sync),
     ) -> Self {
-        assert!(limit > 0, "limit must be a positive integer");
-
         TokenBucket {
-            time_per_token: interval.as_nanos() as usize / limit,
+            time_per_token: if limit > 0 {
+                interval.as_nanos() as usize / limit
+            } else {
+                0
+            },
             interval,
             last_replenished_at: Mutex::new(None),
             clock,
         }
     }
 
+    /// Try to consume the specified number of `tokens` from the bucket.
+    ///
+    /// If the bucket has the sufficient number of tokens available, they are *consumed*
+    /// and `Ok(())` is returned.
+    ///
+    /// If the bucket has fewer tokens available, the internal state is *not* modified,
+    /// and [`Error::RetryAfter`] is returned. The error will specify how much time the
+    /// caller has to wait before trying to call [`TokenBucket::consume()`] with the
+    /// same arguments again. Retrying the operation earlier will result in the same error.
+    ///
+    /// If the bucket has a limit of 0 tokens, [`Error::Blocked`] is always returned instead,
+    /// regardless of how much time the caller waits between attempts.
+    ///
+    /// ```
+    /// use std::time::Duration;
+    /// use youshallnotpass::{TokenBucket, Error};
+    ///
+    /// // create a new bucket that allows to consume 3 tokens every 60 seconds
+    /// let bucket = TokenBucket::new(3, Duration::from_secs(60));
+    /// assert!(bucket.consume(1).is_ok());
+    /// assert!(bucket.consume(1).is_ok());
+    /// assert!(bucket.consume(1).is_ok());
+    /// assert!(matches!(bucket.consume(1), Err(Error::RetryAfter(duration))));
+    ///
+    /// // create a new bucket that does not allow to consume any tokens
+    /// let bucket = TokenBucket::new(0, Duration::from_secs(60));
+    /// assert!(matches!(bucket.consume(1), Err(Error::Blocked)));
+    /// ```
     pub fn consume(&self, tokens: usize) -> Result<(), Error> {
-        let now = (self.clock)();
+        if self.time_per_token == 0 {
+            return Err(Error::Blocked);
+        }
```

Comment:
Q: Makes me wonder, do we want to always pay for comparison on every single `consume()` request whether `time_per_token` is 0 or not? It looks like a waste for most scenarious where `0` is unlikely to explicitly be give. :thinking: 

The only alternative I can think of at the moment is the following:

- Add some `Consumeable` trait
- Add BlockedBucket struct that always return `Blocked`
- Construct proper bucket struct based on the limit in the rate limiter
- Suggest users to use RateLimiter, not TokenBucket (make it private?), to have better UX

But maybe I'm overthinking it. WDYT?

Reply:
That will still not be cost-free, since you'll have to add dynamic dispatch.